### PR TITLE
Low pointer bit macros

### PR DESCRIPTION
**Pull request by qwattash**

### CODE REVIEW

File: sys/kern/kern_lock.c [hunk 1]
```diff
@@ -1462,9 +1475,11 @@ __lockmgr_args(struct lock *lk, u_int flags, struct lock_object *ilk,
 				continue;
 			}
 
-			v = x & (LK_ALL_WAITERS | LK_EXCLUSIVE_SPINNERS);
-			if ((x & ~v) == LK_UNLOCKED) {
-				v = (x & ~LK_EXCLUSIVE_SPINNERS);
+			v = cheri_get_low_ptr_bits(x, LK_ALL_WAITERS |
+			    LK_EXCLUSIVE_SPINNERS);
+			if (cheri_clear_low_ptr_bits(x, v) == LK_UNLOCKED) {
```
Comment (arichardson) [inline sys/kern/kern_lock.c:1480]: Ah I guess this one does need the macros.

**Review by arichardson — COMMENTED** [review]
Are any of these still flagged by the compiler? It looks like most of them should have a clear provenance root and not result in warnings.

I'm not sure whether we should use macros for all cases or try to minimize our diff?

**Comment by jrtc27** [comment]
> Are any of these still flagged by the compiler? It looks like most of them should have a clear provenance root and not result in warnings.
> 
> I'm not sure whether we should use macros for all cases or try to minimize our diff?

Yeah that's just what I was thinking, these shouldn't be needed. The only difference is whether you get a size_t result or a uintcap_t result derived from the pointer, which is only really a problem if you're looking at the low bits of sealed capabilities on !Morello. Even if they are needed, the fact that cheric.h does weird things in hybrid for these (it takes a uintptr_t not uintcap_t) is rather sad.

File: sys/kern/kern_lock.c (hunk 1 above)
Comment (jrtc27) [reply sys/kern/kern_lock.c:1480]: I don't see why?

**Comment by arichardson** [comment]
Using uintcap_t in the macros would unnecessarily promote from long -> uintcap_t. When I wrote those macros, I didn't think about ever using them in hybrid. They were just there to deal with offset mode in purecap and allow non-CHERI to be unchanged.

**Comment by qwattash** [comment]
> Are any of these still flagged by the compiler? It looks like most of them should have a clear provenance root and not result in warnings.
> 
> I'm not sure whether we should use macros for all cases or try to minimize our diff?
I don't think we get things flagged by the compiler at this point, but I'll try reverting and see what compiler warnings I get.
We still have some issue with MIPS codegen though as it emits a candaddr + cbnz
```
// if (v & MTX_FLAG) {}
candaddr $c2, $c2, $at
cbnz $c2, 0x...
```
While RISCV is branching on the address using the corresponding integer register name.

**Comment by jrtc27** [comment]
> > Are any of these still flagged by the compiler? It looks like most of them should have a clear provenance root and not result in warnings.
> > I'm not sure whether we should use macros for all cases or try to minimize our diff?
> 
> We still get some errors from the compiler at this point, because of binary expressions on capabilities where the flags are extracted dynamically from an earlier mtx_lock value.
> We still have some issue with MIPS codegen though as it emits a candaddr + cbnz
> 
> ```
> // if (v & MTX_FLAG) {}
> candaddr $c2, $c2, $at
> cbnz $c2, 0x...
> ```
> 
> While RISCV is branching on the address using the corresponding integer register name.
> I'm happy to retire the macros if they are no longer needed, also keeping this up to date is somewhat annoying when there are changes upstream.

Blegh, CBNZ being the whole capability rather than just the address like CPtrCmp now is is sad, though I guess CPtrCmp with C0 works as an alternative. To be honest, CB[N]Z and CBT[SU] are rather useless with our modern C semantics; both operations should be pretty rare and require the use of intrinsics to generate.

**Comment by arichardson** [comment]
We could probably generate the sensible code (cgetandaddr + bne) with a minor change to codegen to avoid those diffs? Should also avoid the unncessary setaddr for merged register files: https://cheri-compiler-explorer.cl.cam.ac.uk/z/ToYY77

I always though CBNZ only compared the address, and we could change it to do that instead of comparing the full cap to NULL. But I'm not sure it makes any sense to change MIPS at this point.

**Comment by arichardson** [comment]
But of course that only fixes the case where it's used directly in the `if` and doesn't work if you store it to a variable.

**Comment by qwattash** [comment]
> We could probably generate the sensible code (cgetandaddr + bne) with a minor change to codegen to avoid those diffs? Should also avoid the unncessary setaddr for merged register files: https://cheri-compiler-explorer.cl.cam.ac.uk/z/ToYY77
> 
> I always though CBNZ only compared the address, and we could change it to do that instead of comparing the full cap to NULL. But I'm not sure it makes any sense to change MIPS at this point.

I think we should change codegen to use `cgetandaddr`, if we can make these changes unnecesary I think we should.

**Comment by jrtc27** [comment]
> > We could probably generate the sensible code (cgetandaddr + bne) with a minor change to codegen to avoid those diffs? Should also avoid the unncessary setaddr for merged register files: https://cheri-compiler-explorer.cl.cam.ac.uk/z/ToYY77
> > I always though CBNZ only compared the address, and we could change it to do that instead of comparing the full cap to NULL. But I'm not sure it makes any sense to change MIPS at this point.
> 
> I think we should change codegen to use `cgetandaddr`, if we can make these changes unnecesary I think we should.

Eh, in general `uintptr_t & constant` needs to preserve provenance, so you can only do that if the address of the result is the only thing used, which I believe is the case here if you ditch the use of CB[N]Z in the patterns.

**Comment by qwattash** [comment]
> > > We could probably generate the sensible code (cgetandaddr + bne) with a minor change to codegen to avoid those diffs? Should also avoid the unncessary setaddr for merged register files: https://cheri-compiler-explorer.cl.cam.ac.uk/z/ToYY77
> > > I always though CBNZ only compared the address, and we could change it to do that instead of comparing the full cap to NULL. But I'm not sure it makes any sense to change MIPS at this point.
> > 
> > 
> > I think we should change codegen to use `cgetandaddr`, if we can make these changes unnecesary I think we should.
> 
> Eh, in general `uintptr_t & constant` needs to preserve provenance, so you can only do that if the address of the result is the only thing used, which I believe is the case here if you ditch the use of CB[N]Z in the patterns.

Dropping CB[N]Z is also making things consistent with RISC-V and Morello, as there we do address comparison. I'm not sure about whether we want to drop CB[N]Z in all cases, but perhaps in the ones involving `uintptr_t` we should only consider the address.

File: sys/sys/tree.h [hunk 2]
```diff
@@ -348,8 +349,8 @@ struct __no_subobject_bounds {						\
 #define RB_EMPTY(head)			(RB_ROOT(head) == NULL)
 
 #define RB_SET_PARENT(dst, src, field) do {				\
```
Comment (arichardson) [inline sys/sys/tree.h:351]: I'm surprised this hasn't caused issues in userspace.

File: sys/sys/tree.h [hunk 3]
```diff
@@ -33,6 +33,7 @@
 #define	_SYS_TREE_H_
 
 #include <sys/cdefs.h>
+#include "stdint.h"
```
Comment (arichardson) [inline sys/sys/tree.h:36]: <sys/stdint.h>? I don't think any other headers here use relative includes.

**Review by arichardson — APPROVED** [review]

File: sys/sys/tree.h (hunk 2 above)
Comment (qwattash) [reply sys/sys/tree.h:351]: Yeah, not sure how it ended up working, probably it's not used much?

Comment (jrtc27) [reply sys/sys/tree.h:351]: Surprised we don't see sys/sys/rb_test failing...

File: sys/sys/tree.h (hunk 3 above)
Comment (qwattash) [reply sys/sys/tree.h:36]: which file should be used for boostrapping sys/stdint.h in tools/build? I'm not sure where the ptraddr_t typedef should go.

Comment (arichardson) [reply sys/sys/tree.h:36]: Ah I guess we might need a `tools/build/sys/stdint.h` then :(

Comment (jrtc27) [reply sys/sys/tree.h:36]: Or you can use __ptraddr_t

**Pull request merged**